shared/react-native/android/app/src/main/java/io/keybase/ossifrage/KeybasePushNotificationListenerService.java

```diff
@@ -0,0 +1,119 @@
+import java.security.cert.CertificateException;
+import java.util.Map;
+
+import io.keybase.ossifrage.modules.NativeLogger;
+import io.keybase.ossifrage.util.DNSNSFetcher;
+import io.keybase.ossifrage.util.VideoHelper;
+import keybase.Keybase;
+import keybase.PushNotifier;
+
+import static keybase.Keybase.initOnce;
+
+public class KeybasePushNotificationListenerService extends RNPushNotificationListenerService {
+    @Override
+    public void onCreate() {
+        try {
+            Keybase.setGlobalExternalKeyStore(new KeyStore(getApplicationContext(), getSharedPreferences("KeyStore", MODE_PRIVATE)));
+        } catch (KeyStoreException | CertificateException | IOException | NoSuchAlgorithmException e) {
+            NativeLogger.error("Exception in KeybasePushNotificationListenerService.onCreate while trying to link the Android KeyStore to go bind", e);
+        }
+        initOnce(getApplicationContext().getFilesDir().getPath(), "", getApplicationContext().getFileStreamPath("service.log").getAbsolutePath(), "prod", false,
+                new DNSNSFetcher(), new VideoHelper());
+        NativeLogger.info("KeybasePushNotificationListenerService created. path: " + getApplicationContext().getFilesDir().getPath());
+    }
+
+
+    @Override
+    public void onMessageReceived(RemoteMessage message) {
+        NativeLogger.info("KeybasePushNotificationListenerService onMessageReceived");
+        final Bundle bundle = new Bundle();
+        for (Map.Entry<String, String> entry : message.getData().entrySet()) {
+            bundle.putString(entry.getKey(), entry.getValue());
+        }
+        JSONObject data = parseJSONObject(bundle.getString("data"));
+        if (data != null) {
+            if (!bundle.containsKey("message")) {
+                bundle.putString("message", data.optString("alert", null));
+            }
+            if (!bundle.containsKey("title")) {
+                bundle.putString("title", data.optString("title", null));
+            }
+            if (!bundle.containsKey("sound")) {
+                bundle.putString("soundName", data.optString("sound", null));
+            }
+            if (!bundle.containsKey("color")) {
+                bundle.putString("color", data.optString("color", null));
+            }
+
+            final int badge = data.optInt("badge", -1);
+            if (badge >= 0) {
+                ApplicationBadgeHelper.INSTANCE.setApplicationIconBadgeNumber(this, badge);
+            }
+        }
+        try {
+            String type = bundle.getString("type");
+            String convID = bundle.getString("c");
+            String payload = bundle.getString("m");
+            Integer membersType = Integer.parseInt(bundle.getString("t"));
+            PushNotifier notifier = new KBPushNotifier(getApplicationContext());
+            switch (type) {
+                case "chat.newmessageSilent_2": {
+                    Boolean displayPlaintext = "true".equals(bundle.getString("n"));
```

Comment:
What are these keys? Where are they documented? Will they change in the future? Maybe a comment that answers these questions

Reply:
I'm not sure where they are documented, but this is the format that notifications come in from remote. Same as iOS here:
https://github.com/keybase/client/blob/e87639c65b7546c9b618345784ac1989f91c3ba3/shared/react-native/ios/Keybase/AppDelegate.m#L158-L167

Reply:
Also I don't think GSON is useful here, since in the JSON everything is encoded as a string. So we'll have to do all the unruly parsing anyway.

Reply:
@MarcoPolo populated here:
https://github.com/keybase/keybase/blob/3d85540defb35ab0ae4d51a28f46403011632edd/go/gregord/chatpush/push.go#L326-L336